File: services/go-core/analytics/models/metrics.go
package models

import "time"

// DashboardMetrics represents the main dashboard data
type DashboardMetrics struct {
	StartDate         string           `json:"startDate"`
	EndDate           string           `json:"endDate"`
	Overview          OverviewStats    `json:"overview"`
	TopTracks         []TopTrack       `json:"topTracks"`
	ListeningHeatmap  []HeatmapData    `json:"listeningHeatmap"`
	PlaybackTimeline  []TimelineData   `json:"playbackTimeline"`
	GenreDistribution []GenreStats     `json:"genreDistribution"`
	PlatformStats     []PlatformStats  `json:"platformStats"`
	RecentSessions    []SessionSummary `json:"recentSessions"`
	ListeningStreak   StreakData       `json:"listeningStreak"`
}

// OverviewStats represents high-level statistics
type OverviewStats struct {
	TotalPlays              int     `json:"totalPlays"`
	TotalPlayTime           int     `json:"totalPlayTime"`
	AverageSessionDuration  int     `json:"averageSessionDuration"`
	CompletionRate          float64 `json:"completionRate"`
	UniqueTracks            int     `json:"uniqueTracks"`
	UniqueAlbums            int     `json:"uniqueAlbums"`
	UniqueArtists           int     `json:"uniqueArtists"`
	TotalSessions           int     `json:"totalSessions"`
	AverageTracksPerSession float64 `json:"averageTracksPerSession"`
	SkipRate                float64 `json:"skipRate"`
	PlaysChange             float64 `json:"playsChange"`
	PlayTimeChange          float64 `json:"playTimeChange"`
	SessionsChange          float64 `json:"sessionsChange"`
}

// TopTrack represents a top played track
type TopTrack struct {
	TrackID        string  `json:"trackId"`
	TrackTitle     string  `json:"trackTitle"`
	ArtistName     string  `json:"artistName"`
	AlbumTitle     string  `json:"albumTitle"`
	AlbumArt       *string `json:"albumArt,omitempty"`
	PlayCount      int     `json:"playCount"`
	PlayTime       int     `json:"playTime"`
	CompletionRate float64 `json:"completionRate"`
	LastPlayed     string  `json:"lastPlayed"`
	Rank           int     `json:"rank"`
	RankChange     int     `json:"rankChange"`
}

// HeatmapData represents a single heatmap data point
type HeatmapData struct {
	Date      string `json:"date"`
	Value     int    `json:"value"`
	DayOfWeek int    `json:"dayOfWeek"`
	Hour      *int   `json:"hour,omitempty"`
}

// TimelineData represents playback timeline data
type TimelineData struct {
	Timestamp              string  `json:"timestamp"`
	Date                   string  `json:"date"`
	PlayCount              int     `json:"playCount"`
	PlayTime               int     `json:"playTime"`
	UniqueTracks           int     `json:"uniqueTracks"`
	UniqueArtists          int     `json:"uniqueArtists"`
	AverageSessionDuration int     `json:"averageSessionDuration"`
	CompletionRate         float64 `json:"completionRate"`
}

// GenreStats represents genre distribution statistics
type GenreStats struct {
	Genre      string     `json:"genre"`
	PlayCount  int        `json:"playCount"`
	PlayTime   int        `json:"playTime"`
	TrackCount int        `json:"trackCount"`
	Percentage float64    `json:"percentage"`
	TopArtist  *TopArtist `json:"topArtist,omitempty"`
}

// TopArtist represents the top artist in a genre
type TopArtist struct {
	Name      string `json:"name"`
	PlayCount int    `json:"playCount"`
}

// PlatformStats represents platform usage statistics
type PlatformStats struct {
	Platform       string  `json:"platform"`
	Browser        *string `json:"browser,omitempty"`
	BrowserVersion *string `json:"browserVersion,omitempty"`
	OS             *string `json:"os,omitempty"`
	OSVersion      *string `json:"osVersion,omitempty"`
	SessionCount   int     `json:"sessionCount"`
	PlayCount      int     `json:"playCount"`
	PlayTime       int     `json:"playTime"`
	Percentage     float64 `json:"percentage"`
	LastUsed       string  `json:"lastUsed"`
}

// SessionSummary represents a session summary
type SessionSummary struct {
	SessionID     string    `json:"sessionId"`
	StartTime     string    `json:"startTime"`
	EndTime       *string   `json:"endTime,omitempty"`
	Duration      int       `json:"duration"`
	Platform      string    `json:"platform"`
	Browser       *string   `json:"browser,omitempty"`
	TracksPlayed  int       `json:"tracksPlayed"`
	UniqueArtists int       `json:"uniqueArtists"`
	TotalPlayTime int       `json:"totalPlayTime"`
	TopTrack      *TopTrack `json:"topTrack,omitempty"`
}

// StreakData represents listening streak information
type StreakData struct {
	CurrentStreak    int        `json:"currentStreak"`
	LongestStreak    int        `json:"longestStreak"`
	StreakStartDate  string     `json:"streakStartDate"`
	TotalDaysActive  int        `json:"totalDaysActive"`
	TotalWeeksActive int        `json:"totalWeeksActive"`
	NextMilestone    *Milestone `json:"nextMilestone,omitempty"`
}

// Milestone represents a streak milestone
type Milestone struct {
	Days      int `json:"days"`
	DaysUntil int `json:"daysUntil"`
}

// TrackSegmentAnalysis represents detailed segment analysis for a track
type TrackSegmentAnalysis struct {
	TrackID           string             `json:"trackId"`
	TrackTitle        string             `json:"trackTitle"`
	ArtistName        string             `json:"artistName"`
	Duration          int                `json:"duration"`
	Segments          []TrackSegmentData `json:"segments"`
	MostPlayedSegment MostPlayedSegment  `json:"mostPlayedSegment"`
	SkipPoints        []SkipPoint        `json:"skipPoints"`
}

// TrackSegmentData represents a single segment
type TrackSegmentData struct {
	TrackID       string  `json:"trackId"`
	SegmentStart  int     `json:"segmentStart"`
	SegmentEnd    int     `json:"segmentEnd"`
	PlayCount     int     `json:"playCount"`
	HeatIntensity float64 `json:"heatIntensity"`
}

// MostPlayedSegment represents the most played segment
type MostPlayedSegment struct {
	Start     int `json:"start"`
	End       int `json:"end"`
	PlayCount int `json:"playCount"`
}

// SkipPoint represents a position where users skip
type SkipPoint struct {
	Position  int `json:"position"`
	SkipCount int `json:"skipCount"`
}

// ListeningPattern represents user listening patterns
type ListeningPattern struct {
	PeakListeningHour    int      `json:"peakListeningHour"`
	PeakListeningDay     int      `json:"peakListeningDay"`
	AverageSessionLength int      `json:"averageSessionLength"`
	PreferredGenres      []string `json:"preferredGenres"`
	SkipBehavior         string   `json:"skipBehavior"`
	NewArtistsPerWeek    int      `json:"newArtistsPerWeek"`
	RepeatListeningRate  float64  `json:"repeatListeningRate"`
	PreferredCodecs      []string `json:"preferredCodecs"`
	AverageBitrate       int      `json:"averageBitrate"`
}

// QueryFilters represents filters for analytics queries
type QueryFilters struct {
	StartDate *time.Time `json:"startDate,omitempty"`
	EndDate   *time.Time `json:"endDate,omitempty"`
	Period    *string    `json:"period,omitempty"` // day, week, month, year, all
	Platform  *string    `json:"platform,omitempty"`
	Genre     *string    `json:"genre,omitempty"`
	ArtistID  *string    `json:"artistId,omitempty"`
	AlbumID   *string    `json:"albumId,omitempty"`
	GroupBy   *string    `json:"groupBy,omitempty"` // hour, day, week, month
	Limit     int        `json:"limit,omitempty"`
	Offset    int        `json:"offset,omitempty"`
}

File: packages/analytics/src/types/metrics.ts
/**
 * Analytics Metrics Types
 * 
 * Defines the structure of aggregated metrics and dashboard data.
 * These are the insights derived from raw events.
 */

// ============================================================================
// Dashboard Overview
// ============================================================================

/**
 * Main dashboard data structure
 */
export interface DashboardMetrics {
  // Time Range
  startDate: string;
  endDate: string;
  
  // Overview Stats
  overview: OverviewStats;
  
  // Charts Data
  topTracks: TopTrack[];
  listeningHeatmap: HeatmapData[];
  playbackTimeline: TimelineData[];
  genreDistribution: GenreStats[];
  platformStats: PlatformStats[];

  // Top Lists
  topArtists: TopArtist[];
  topAlbums: TopAlbum[];
  topPlaylists: TopPlaylist[];

  // Recent Activity
  recentSessions: SessionSummary[];
  recentlyPlayed: RecentlyPlayedTrack[];
  listeningStreak: StreakData;
}

/**
 * Artist specific metrics
 */
export interface ArtistMetrics {
  artistId?: string;
  artistName: string;
  overview: OverviewStats;
  topTracks: TopTrack[];
  playbackTimeline: TimelineData[];
  fanDemographics: PlatformStats[];
}

/**
 * Album specific metrics
 */
export interface AlbumMetrics {
  albumId?: string;
  albumTitle: string;
  artistName: string;
  overview: OverviewStats;
  trackPerformance: TopTrack[];
  playbackTimeline: TimelineData[];
}

// ============================================================================
// Overview Statistics
// ============================================================================

export interface OverviewStats {
  // Playback Metrics
  totalPlays: number;
  totalPlayTime: number; // seconds
  averageSessionDuration: number; // seconds
  completionRate: number; // percentage
  
  // Library Metrics
  uniqueTracks: number;
  uniqueAlbums: number;
  uniqueArtists: number;
  
  // Engagement Metrics
  totalSessions: number;
  averageTracksPerSession: number;
  skipRate: number; // percentage
  
  // Comparisons (vs previous period)
  playsChange: number; // percentage
  playTimeChange: number; // percentage
  sessionsChange: number; // percentage
  
  // Library Stats
  totalTracksInLibrary: number;
  totalAlbumsInLibrary: number;
  totalArtistsInLibrary: number;
  formats: FormatStat[];
  coverArtStats: CoverArtSummary;
}

export interface FormatStat {
  format: string;
  count: number;
  size?: string;
}

export interface CoverArtSummary {
  withCover: number;
  withoutCover: number;
  percentage: number;
}

// ============================================================================
// Top Tracks
// ============================================================================

export interface TopTrack {
  trackId: string;
  trackTitle: string;
  artistName: string;
  albumTitle: string;
  albumArt?: string;
  
  playCount: number;
  playTime: number; // total seconds
  completionRate: number; // percentage
  lastPlayed: string; // ISO timestamp
  
  rank: number;
  rankChange: number; // vs previous period
}

export interface TopArtist {
  artistId: string;
  artistName: string;
  playCount: number;
  playTime: number; // total seconds
  rank: number;
}

export interface TopAlbum {
  albumId: string;
  albumTitle: string;
  artistName: string;
  coverArt?: string;
  playCount: number;
  playTime: number; // total seconds
  rank: number;
}

export interface TopPlaylist {
  playlistId: string;
  playlistName: string;
  playCount: number;
  playTime: number; // total seconds
  rank: number;
}

export interface RecentlyPlayedTrack {
  trackId: string;
  trackTitle: string;
  artistName: string;
  albumTitle: string;
  albumArt?: string;
  playedAt: string; // ISO timestamp
  duration?: number; // seconds
}

// ============================================================================
// Listening Heatmap
// ============================================================================

export interface HeatmapData {
  date: string; // YYYY-MM-DD
  value: number; // play count or duration
  dayOfWeek: number; // 0-6
  hour?: number; // 0-23 for hourly heatmap
}

export interface HourlyHeatmap {
  hour: number; // 0-23
  dayOfWeek: number; // 0-6
  playCount: number;
  averageDuration: number;
}

// ============================================================================
// Playback Timeline
// ============================================================================

export interface TimelineData {
  timestamp: string; // ISO timestamp
  date: string; // YYYY-MM-DD
  
  // Metrics
  playCount: number;
  playTime: number; // seconds
  uniqueTracks: number;
  uniqueArtists: number;
  
  // Averages
  averageSessionDuration: number;
  completionRate: number;
}

// ============================================================================
// Genre Distribution
// ============================================================================

export interface GenreStats {
  genre: string;
  playCount: number;
  playTime: number; // seconds
  trackCount: number;
  percentage: number; // of total plays
  
  // Top artist in this genre
  topArtist?: {
    name: string;
    playCount: number;
  };
}

// ============================================================================
// Platform Statistics
// ============================================================================

export interface PlatformStats {
  platform: string; // 'web', 'mobile', 'desktop'
  browser?: string;
  browserVersion?: string;
  os?: string;
  osVersion?: string;
  
  sessionCount: number;
  playCount: number;
  playTime: number; // seconds
  percentage: number; // of total sessions
  
  lastUsed: string; // ISO timestamp
}

// ============================================================================
// Session Summary
// ============================================================================

export interface SessionSummary {
  sessionId: string;
  startTime: string; // ISO timestamp
  endTime?: string; // ISO timestamp
  duration: number; // seconds
  
  platform: string;
  browser?: string;
  
  // Activity
  tracksPlayed: number;
  uniqueArtists: number;
  totalPlayTime: number; // seconds
  
  // Most played in session
  topTrack?: {
    trackId: string;
    title: string;
    artist: string;
    playCount: number;
  };
}

// ============================================================================
// Listening Streak
// ============================================================================

export interface StreakData {
  currentStreak: number; // days
  longestStreak: number; // days
  streakStartDate: string; // ISO timestamp
  
  // Milestones
  totalDaysActive: number;
  totalWeeksActive: number;
  
  // Predictions
  nextMilestone?: {
    days: number;
    daysUntil: number;
  };
}

// ============================================================================
// Track Segments (Most Listened Parts)
// ============================================================================

export interface TrackSegment {
  trackId: string;
  segmentStart: number; // seconds
  segmentEnd: number; // seconds
  playCount: number;
  heatIntensity: number; // 0-1, normalized
}

export interface TrackSegmentAnalysis {
  trackId: string;
  trackTitle: string;
  artistName: string;
  duration: number;
  
  segments: TrackSegment[];
  
  // Insights
  mostPlayedSegment: {
    start: number;
    end: number;
    playCount: number;
  };
  
  skipPoints: {
    position: number; // seconds
    skipCount: number;
  }[];
}

// ============================================================================
// Listening Patterns
// ============================================================================

export interface ListeningPattern {
  // Time-based patterns
  peakListeningHour: number; // 0-23
  peakListeningDay: number; // 0-6
  
  // Behavior patterns
  averageSessionLength: number; // seconds
  preferredGenres: string[];
  skipBehavior: 'patient' | 'moderate' | 'impatient';
  
  // Discovery patterns
  newArtistsPerWeek: number;
  repeatListeningRate: number; // percentage
  
  // Quality preferences
  preferredCodecs: string[];
  averageBitrate: number;
}

// ============================================================================
// Query Filters
// ============================================================================

export interface AnalyticsFilters {
  // Time Range
  startDate?: string;
  endDate?: string;
  period?: 'day' | 'week' | 'month' | 'year' | 'all';
  
  // Filters
  platform?: string;
  genre?: string;
  artistId?: string;
  albumId?: string;
  artistName?: string;
  albumTitle?: string;
  
  // Aggregation
  groupBy?: 'hour' | 'day' | 'week' | 'month';
  
  // Pagination
  limit?: number;
  offset?: number;
}

// ============================================================================
// Export Data
// ============================================================================

export interface AnalyticsExport {
  exportDate: string;
  dateRange: {
    start: string;
    end: string;
  };
  
  // Data
  sessions: SessionSummary[];
  topTracks: TopTrack[];
  listeningPatterns: ListeningPattern;
  
  // Metadata
  totalEvents: number;
  exportFormat: 'json' | 'csv';
}
